src/codegen.cpp

```diff
@@ -1382,367 +1048,101 @@ static void jl_setup_module(Module *m, const jl_cgparams_t *params = &jl_default
+            needsparams = true;
+        for (size_t i = 0; i < jl_svec_len(lam->sparam_vals); ++i) {
+            if (jl_is_typevar(jl_svecref(lam->sparam_vals, i)))
+                needsparams = true;
         }
     }
-    if (src == NULL || (jl_value_t*)src == jl_nothing || !jl_is_code_info(src))
-        jl_error("source not available for function");
-
-    // Backup the info for the nested compile
-    JL_LOCK(&codegen_lock);
-
-    // emit this function into a new module
-    jl_llvm_functions_t declarations;
-    std::unique_ptr<Module> m;
-    JL_TRY {
-        m = emit_function(mi, src, jlrettype, world, &declarations, &params);
-    }
-    JL_CATCH {
-        // something failed!
-        m.reset();
-        JL_UNLOCK(&codegen_lock); // Might GC
-        const char *mname = name_from_method_instance(mi);
-        jl_rethrow_with_add("error compiling %s", mname);
-    }
-
-    if (optimize)
-        jl_globalPM->run(*m.get());
-
-    // swap declarations for definitions and destroy declarations
-    const char *fname = declarations.functionObject;
-    const char *specfname = declarations.specFunctionObject;
-    Function *f = NULL;
-    Function *specf = NULL;
-    if (specfname) {
-        specf = cast<Function>(m->getNamedValue(specfname));
-        free(const_cast<char*>(specfname));
-    }
-    f = cast_or_null<Function>(m->getNamedValue(fname));
-    if (f) // don't try to free sentinel names like "jl_fptr_args" and "jl_fptr_sparam"
-        free(const_cast<char*>(fname));
-    assert(specf || f);
-    // try to clone the name from some linfo, if it exists
-    // to give the user a (false) sense of stability
-    jl_code_instance_t *codeinst = mi->cache;
-    while (codeinst) {
-        specfname = codeinst->functionObjectsDecls.specFunctionObject;
-        if (specfname && specf) {
-            specf->setName(specfname);
-        }
-        fname = codeinst->functionObjectsDecls.functionObject;
-        if (fname && f) {
-            if (strcmp(fname, "jl_fptr_args") && strcmp(fname, "jl_fptr_sparam"))
-                f->setName(fname);
-            break;
-        }
-        codeinst = codeinst->next;
+    if (needsparams)
+        return std::make_pair(false, true);
+    if (sig == (jl_value_t*)jl_anytuple_type)
+        return std::make_pair(false, false);
+    if (!jl_is_datatype(sig))
+        return std::make_pair(false, false);
+    if (jl_nparams(sig) == 0)
+        return std::make_pair(false, false);
+    if (va) {
+        if (jl_is_vararg_type(jl_tparam(sig, jl_nparams(sig) - 1)))
+            return std::make_pair(false, false);
     }
-    m.release(); // the return object `llvmf` will be the owning pointer
-    JL_UNLOCK(&codegen_lock); // Might GC
-    JL_GC_POP();
-    if ((getwrapper && f) || !specf)
-        return f;
-    else
-        return specf;
-}
-
-
-extern "C" JL_DLLEXPORT
-void *jl_get_llvmf_decl(jl_method_instance_t *mi, size_t world, bool getwrapper, const jl_cgparams_t params)
```

Comment:
Is there an alternative function to use instead of this one? It was used in Cxx.jl: https://github.com/JuliaInterop/Cxx.jl/blob/256485c04c76feee515e83f091ba0d6c8f3286ee/src/cxxstr.jl#L107